### ISSUE TITLE
[BUG] Very slow Generation with Paged Attention

### ISSUE THREAD

**Issue by rjmehta1993** [turboderp/exllamav2#679]
### OS

Linux

### GPU Library

CUDA 12.x

### Python version

3.10

### Pytorch version

2.5.0

### Model

DrNicefellow/Qwen2.5-32B-Instruct-8.0bpw-h8-exl2

### Describe the bug

If I understand correctly, paged attention allows multiple requests to be decoded using KV cache blocks. When I process a single request, it takes ~30 seconds. Running 10 requests one by one takes 5 minutes. 

Question: When I run 10 requests using paged attention, it takes ~3 mins to finish. With batch_size as 10, shouldn't that finish in <1 min if each batch is running in parallel?

I am using Qwen2.5 32B 8bpw with rope scaling and 128k context len. I am using A10G 4X24GB GPU

GPU Memory: 96Gb

Also, if I use a 7b draft model, it takes even longer than 3 mins.

Also, I am using Cuda 12.4 with flash attn 2.6.3 and the latest exl version.

Questions: 
1. Why is the generation slow for each request when using paged attention and multiple requests in parallel?
2. Why does the generation slows when using draft model?
3. If using TP, does the generation throughput increase?
4. What could be the best settings to have higher throughput?

### Reproduction steps

I am using async with paged attention using the below config:

SERVER: 
```
LLM_CONTEXT_LEN = 128000

# FastAPI app
app = FastAPI()

# Define global variables for model, cache, tokenizer, and generator
model = None
cache = None
tokenizer = None
generator = None
use_ngram = False
max_input_len = 8192
max_attention_size = max_input_len ** 2
total_cache_across_jobs = 128000
batch_size = 100

use_draft_model = None

@app.on_event("startup")
async def startup_event():
      
    config = ExLlamaV2Config(AI_MODEL)

    config.max_seq_len = LLM_CONTEXT_LEN
    config.max_input_len = max_chunk_size
    config.max_attention_size = max_attention_size
    model = ExLlamaV2(config)
    cache = ExLlamaV2Cache(model, max_seq_len = total_cache_across_jobs, lazy=True)
    model.load_autosplit(cache, progress=True)

    logger.info("Loading tokenizer...")
    tokenizer = ExLlamaV2Tokenizer(config)
    generator = ExLlamaV2DynamicGeneratorAsync(
        model=model,
        cache=cache,
        tokenizer=tokenizer,
        max_batch_size=batch_size,
        max_chunk_size=max_chunk_size,
        paged=True,
    )
```

CLIENT:
```

from concurrent.futures import ThreadPoolExecutor, as_completed

# Function to execute multiple requests concurrently
def execute_requests_model_pred():
    updated_res = []
    instr = "You are an assistant"
    user = """<Document Start>\n{8K input document}\n<Document End>\n\nQuestion: """

    with ThreadPoolExecutor(max_workers=batchsize) as executor:
        futures = []
        for i in range(len(df)):
            Question = df.loc[i, "Question"]
            Context = df.loc[i, "Context"]
            user = """<Document Start>\n{8K input document}\n<Document End>\n\nQuestion: """
            futures.append(executor.submit(run_llm, ins = instr, \
                                   user = user))
            
        for future in tqdm(as_completed(futures), total=len(futures), desc="Requests Completed"):
            updated_res.append(future.result())
    return updated_res

```

### Expected behavior

The time to finish 10 batch_size with paged attention should be less than 1 min.

### Acknowledgements

- [X] I have looked for similar issues before submitting this one.
- [X] I understand that the developers have lives and my issue will be answered when possible.
- [X] I understand the developers of this program are human, and I will ask my questions politely.

**Comment by rjmehta1993** [comment]
@turboderp If you can help with this one whenever you find time. Thanks :)

**Comment by turboderp** [comment]
I would need to know a little bit more about the requests. If each request has a very long, unique prompt, then inference is already batching 8192 tokens tokens together (one chunk as defined by `max_input_len`) and you can't parallelize any more than that because inference will be heavily compute-limited.

Another question would be what `max_new_tokens` is set to for each job. If each needs to reserve 64k tokens of the cache, you'll more or less be running 5 batches of batch size 2.

Using a draft model can increase performance in some cases, but it does so by batching multiple tokens at once for a single sequence and hoping they were predicted correctly. If you're already batching many jobs together there isn't that much more to gain from speculative decoding, but if the draft model isn't decreasing accurately you may still get the penalties from that.

Also a 7B draft model is generally too large for a 32B model, even at batch size 1, unless you're expecting *really* good predictions from the 7B model.

And TP can improve performance quite a bit, yes, but it is also a little fickle. It works mostly for the generation stage, while prompt ingestion suffers due to the PCIe overhead.

What results do you get from running the `bulk_inference.py` example?

**Comment by rjmehta1993** [comment]
1. Each request is around 8k unique tokens and unique prompts. Prompt caching wont help here.
2. max_new_tokens is around 512 tokens at max.

- Yeah draft model is penalizing the throughput but not the accuracy. I used 7b as a draft, and 1.5 b has a different vocabulary size. 7b is the smallest that matches 32b for qwen 2.5

- Yes, I tried TP, and it improved the throughput, but the accuracy dropped from 83 to 68. That's a substantial drop.

- My inference is more like adding a request in real time as opposed to a predefined list of inferences like in bulk_inference.py. Hence, I used thread to create 32, 64, and 100 threads each being sent to the model as async.

- Does bulk_inference.py compile 100 requests at once as a list and send them to the model as opposed to threads where each request is sent one by one?

**Comment by rjmehta1993** [comment]
@turboderp Your help is appreciated :)

**Comment by remichu-ai** [comment]
I think you should be using async generator instead of non async generator with thread. Async generator will batch the generation automatically.

If u have more gpu than the model itself and want to increase the throughtput, you can load additional model in another fastapi endpoint or using the same endpoint but using multiprocessing.

Try to load multiple copy of the models in the same fastapi wont increase speed without multiprocessing due to python and fastapi have single thread constraints i believe.

I load multiple models in different endpoints and that work for me. U can refer to sample in repo under my id.

**Comment by turboderp** [comment]
> 1. Each request is around 8k unique tokens and unique prompts. Prompt caching wont help here.
> 
> 2. max_new_tokens is around 512 tokens at max.

30 seconds is a lot for a single request in this case. I'm not sure about the A10G, though. It's not one of the GPUs I've specifically had in mind when optimizing. Overall performance should be something like 60% that of the 3090, but I'm not sure how that manifests in relation to scaling an inference batch. There is a possibility that you're simply only getting about a 2x increase in total tokens/second with batching. But with 10x8.5k tokens you should be fitting everything in the cache, at least.
 
>     * Yeah draft model is penalizing the throughput but not the accuracy. I used 7b as a draft, and 1.5 b has a different vocabulary size. 7b is the smallest that matches 32b for qwen 2.5

Are you sure it won't work though? 0.5B and 1.5B have slightly smaller vocab sizes defined in their configs, but looking at the respective tokenizers they do look like they're identical vocabularies. The difference appears to be just padding, which wouldn't matter as long as the draft model has less of it.
 
>     * Yes, I tried TP, and it improved the throughput, but the accuracy dropped from 83 to 68. That's a substantial drop.

This is definitely some sort of separate issue. TP should not affect the output quality. I'll run some more tests.
 
>     * My inference is more like adding a request in real time as opposed to a predefined list of inferences like in bulk_inference.py. Hence, I used thread to create 32, 64, and 100 threads each being sent to the model as async.
 
Adding requests in real time with the async generator is fine, but I'm not sure how it mixes with multithreading. There might be something funny about how iterations are getting scheduled since an async job really expects to be run in a coroutine.

>     * Does bulk_inference.py compile 100 requests at once as a list and send them to the model as opposed to threads where each request is sent one by one?

It creates 10,000 jobs and schedules them all at once. Scheduling is sequential, though, and there's no restriction on when a job can be added to the queue. On each iteration the generator pulls as many pending jobs into the active list as it can, then generates one token from all the active jobs. If a job completes, it's evicted from the active list and on the next iteration another pending job will be activated, if possible.
